registry/handlers/app.go

```diff
@@ -232,6 +227,83 @@ func (app *App) register(routeName string, dispatch dispatchFunc) {
 	app.router.GetRoute(routeName).Handler(app.dispatcher(dispatch))
 }
 
+func (app *App) configureDriver(configuration *configuration.Configuration) {
+	for {
+		var err error
+		app.driver, err = factory.Create(configuration.Storage.Type(), configuration.Storage.Parameters())
+		if err != nil {
+			// TODO(stevvooe): Move the creation of a service into a protected
+			// method, where this is created lazily. Its status can be queried via
+			// a health check.
+			panic(err)
+		}
+
+		// HACK(stevvooe): Previously, the storage package automatically added
+		// "/docker/registry" as a path prefix for all registry paths. The
+		// intention was to have common root configuration for all drivers.
+		// Instead, we've opted to leave this up to the driver by allowing the
+		// configuration of "rootdirectory" or similar configuration.
+		//
+		// To make the transition smoother, we are adding a nice hack to
+		// rewrite the rootdirectory path for the drivers that support it. To
+		// do this, we issue a few listings, rewrite the configuration, if
+		// necessary, then loop again.
+		//
+		// For all drivers added after 2.1 release, this doesn't matter.
+		//
+		// Leave this detection in for 2.2 to provide people time to update
+		// their configurations. The below can be removed in 2.3.
+
+		// detect a fresh registry: empty root. "/v2" and "/docker/registry"
+		// don't exist.
+		entries, err := app.driver.List(app, "/")
+		if err != nil {
+			panic(err) // what should be done here?
```

Comment:
This is matching the current behavior. We'll have to fix this but we need to provide actionable feedback on misconfigurations. This is often the best way.

Processes should be restarted on failure, anyways.